wire/msgmixcm.go

```diff
@@ -0,0 +1,212 @@
+		msg := fmt.Sprintf("too many previous referenced messages [%v]", count)
+		return messageError(op, ErrTooManyPrevMixMsgs, msg)
+	}
+
+	seen := make([]chainhash.Hash, count)
+	for i := range seen {
+		err := readElement(r, &seen[i])
+		if err != nil {
+			return err
+		}
+	}
+	msg.SeenDCs = seen
+
+	return nil
+}
+
+// BtcEncode encodes the receiver to w using the Decred protocol encoding.
+// This is part of the Message interface implementation.
+func (msg *MsgMixCM) BtcEncode(w io.Writer, pver uint32) error {
+	const op = "MsgMixCM.BtcEncode"
+	if pver < MixVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	err := writeElement(w, &msg.Signature)
+	if err != nil {
+		return err
+	}
+
+	err = msg.writeMessageNoSignature(op, w, pver)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeMessageNoSignature serializes all elements of the message except for
+// the signature.  This allows code reuse between message serialization, and
+// signing and verifying these message contents.
+func (msg *MsgMixCM) writeMessageNoSignature(op string, w io.Writer, pver uint32) error {
+	err := writeElements(w, &msg.Identity, &msg.SessionID, msg.Expiry,
+		msg.Run)
+	if err != nil {
+		return err
+	}
+
+	err = msg.Mix.BtcEncode(w, pver)
+	if err != nil {
+		return err
+	}
+
+	count := len(msg.SeenDCs)
+	if count > MaxPrevMixMsgs {
+		msg := fmt.Sprintf("too many previous referenced messages [%v]", count)
+		return messageError(op, ErrTooManyPrevMixMsgs, msg)
+	}
+
+	err = WriteVarInt(w, pver, uint64(count))
+	if err != nil {
+		return err
+	}
+	for i := range msg.SeenDCs {
+		err = writeElement(w, &msg.SeenDCs[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WriteSigned writes a tag identifying the message data, followed by all
+// message fields excluding the signature.  This is the data committed to when
+// the message is signed.
+func (msg *MsgMixCM) WriteSigned(w io.Writer) error {
+	const op = "MsgMixCM.WriteSigned"
+
+	err := WriteVarString(w, MixVersion, CmdMixCM+"-sig")
+	if err != nil {
+		return err
+	}
+
+	err = msg.writeMessageNoSignature(op, w, MixVersion)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Command returns the protocol command string for the message.  This is part
+// of the Message interface implementation.
+func (msg *MsgMixCM) Command() string {
+	return CmdMixCM
+}
+
+// MaxPayloadLength returns the maximum length the payload can be for the
+// receiver.  This is part of the Message interface implementation.
+func (msg *MsgMixCM) MaxPayloadLength(pver uint32) uint32 {
+	return 0 // XXX 32 + 32 + MaxTxSize + 4 + 64
+}
+
+// Hash returns the hash of the serialized message.
+func (msg *MsgMixCM) Hash() chainhash.Hash {
+	return mustHash(msg, MixVersion)
+}
+
+// GetIdentity returns the message sender's public key identity.
+func (msg *MsgMixCM) GetIdentity() []byte {
```

Comment:
Yes, I wasn't in love with the name.  I would have preferred Identity, but it was already a field, and could not also be used as a method.  Ident as an alternative would be ok (either in the method name, or the field name).

Not sure what you mean in that last sentence about hashing.  This field is for a secp256k1 public key.

Reply:
If you haven't gotten to the mixing module review yet, it defines the interface type that we are implementing here in wire.  This means we can do stuff like read a wire message and check if it is able to be type asserted to a mixing.Message.

The alternative is to explicitly check the concrete types of the wire.Message and wrap it with some other concrete mixing types (a la dcrutil), but that feels like unnecessary busy work to juggle the two different types around.

Reply:
If the comment about hashing is because wrapper types in the mixing package would implement Hash() and WriteSigned(), that's not so easy to do without a massive amount of code duplication.  We would need to either export writeMessageNoSignature, or consistently create a copy and zero it out before writing the entire message plus tag.  Otherwise, there is a lot of unexported wire helper functions that would need to be copied into the mixing package as well.